Implement 2-cohomology and module computations for arbitrary finite groups, not just solvable ones, via `TwoCohomologyGeneric`

hulpke opened the pull request:
Uses rewriting system since pc presentation will not always be available.
This is both code for the cohomology group, and for constructing the extension corresponding to a particular cocycle. Group is constructed by default as Fp group, but there is an option to have it compute a faithful permutation representation.

Also includes improved routines for `IrreducibleModules` for the non-pc case once such cohomology routines exist, and action of compatible pairs on the cohomology group.

There is a test in `testextra` that, as application, constructs perfect groups anew. (This could easily go to larger orders, if desired, at the moment two quick tests are in.)

Commit:
ENHANCE: Two-Cohomology for arbitrary finite groups

Uses rewriting system since pc presentation will not always be available.

Use a rewriting system based on extensions for the factor group (as it is
smaller).
Rename `PresentationCocycle` to `FpGroupCocycle` and offer to calculate a
faithful permutation representation for it.

Added test

Add info from decomp.

Avoid the global Gpword fct. for order issues.

ENHANCE: Allow for different presentations for better rewriting

Added IsomorphismFpGroupForRewriting and method for alternating groups.

ENHANCE: Avoid duplicate test, build on existing permrep first.

Examples reflect new code. Make immutable.

Simplify presentation before trying permrep.

Force natural An for presentation.

REBASE: Fix wrong suggestion.

ENHANCE: Two-Cohomology for arbitrary finite groups

Uses rewriting system since pc presentation will not always be available.

Use a rewriting system based on extensions for the factor group (as it is
smaller).
Rename `PresentationCocycle` to `FpGroupCocycle` and offer to calculate a
faithful permutation representation for it.

Added test

Add info from decomp.

Avoid the global Gpword fct. for order issues.

ENHANCE: Allow for different presentations for better rewriting

Added IsomorphismFpGroupForRewriting and method for alternating groups.

ENHANCE: Avoid duplicate test, build on existing permrep first.

Examples reflect new code. Make immutable.

Simplify presentation before trying permrep.

Force natural An for presentation.

Simplify presentation before finding subgroup quotients

Run Tietze in LargerAbelianQuotient with different heuristics

ENHANCE: Two-Cohomology for arbitrary finite groups

Uses rewriting system since pc presentation will not always be available.

Use a rewriting system based on extensions for the factor group (as it is
smaller).
Rename `PresentationCocycle` to `FpGroupCocycle` and offer to calculate a
faithful permutation representation for it.

Added test

Add info from decomp.

Avoid the global Gpword fct. for order issues.

ENHANCE: Allow for different presentations for better rewriting

Added IsomorphismFpGroupForRewriting and method for alternating groups.

ENHANCE: Avoid duplicate test, build on existing permrep first.

Examples reflect new code. Make immutable.

Simplify presentation before trying permrep.

Force natural An for presentation.

REBASE: Fix wrong suggestion.

REBASE: Simplify presentation before finding subgroup quotients

Run Tietze in LargerAbelianQuotient with different heuristics

fingolfin commented on doc/ref/grppc.xml:
```diff
@@ -417,6 +417,8 @@ The only method with is special for pc groups is a method
 to compute intersections of subgroups, since here a pcgs of a parent
 group is needed and this can only by guaranteed within pc groups.
 
+Section <Ref Sect="2-Cohomology"> describes operations and methods for
```
```suggestion
Section <Ref Sect="2-Cohomology"/> describes operations and methods for
```

hulpke replied:
Thanks.

Commit:
ENHANCE: Homomorphisms for 1-dimensional modules just is common eigenspace.

Commit:
TWOCOHOM: Added tests: Construction of perfect groups

Includes action of compatible pairs on cohomology group and model.

wilfwilson commented:
@hulpke I noticed you’ve removed the WIP from the title and the ‘do not review’ label. Just to check, for sure: does this mean that you’d be happy for the PR to be reviewed, and hopefully merged? If so, I’d suggest removing the ‘do not merge’ label because I get the impression it might put people off from reviewing. Sorry I’ve not had chance to look at your PR further yet.

alex-konovalov commented:
Travis reports two manual example diffs:
```
########> Diff in /home/travis/build/gap-system/gap/doc/ref/../../lib/twocohom\
.gd:273 (/home/travis/build/gap-system/gap/tst/testmanuals/ref-chapter039.tst:
992)
# Input is:
g1:=Image(IsomorphismPermGroup(p));
# Expected output:
<permutation group of size 38880 with 9 generators>
# But found:
<permutation group with 9 generators>
########
########> Diff in /home/travis/build/gap-system/gap/doc/ref/../../lib/twocohom\
.gd:273 (/home/travis/build/gap-system/gap/tst/testmanuals/ref-chapter039.tst:
999)
# Input is:
g2:=Image(IsomorphismPermGroup(p));
# Expected output:
<permutation group of size 38880 with 9 generators>
# But found:
<permutation group with 9 generators>
########
```

hulpke commented:
@wilfwilson 
At this stage I am happy for comments about potential problems, but there is still some code (basic functionality for compatible pairs and action on the cohomology group) I am still working on and want to add before merging. (There also is a manual bug and a stupid typo in the perfect group construction example that need fixing.)
As there will still be code added (as soon as I fix a bug..), it might be best to wait with formal reviewing.

hulpke commented:
@wilfwilson OK, now the promised changes are in, and the holds are removed.
Since there is no need to backport, I'm not sure whether there is a need to squash into one commit, but could do so if desired.

fingolfin reviewed (changes requested):
Very nice, thank you! Overall, this is of course a very welcome improvement!

Unfortunately I did not yet have time to look at everything (with easter and everything), but I left a few rather minor remarks. I would like to have a closer look at a few more things, I'll try to get this done this week or early next week.

Of course it would be nice to have this in GAP 4.11 -- @alex-konovalov is in charge of that.

hulpke commented on lib/grpreps.gi:
```diff
@@ -44,21 +44,194 @@ end);
+
+    # get augmentation ideal of subgroup, at least approximate from
+    # generators
+    a:=AsList(G);
```
Will change so (but it anyhow is a fallback method, permutation groups use Burnside/Brauer method below).

hulpke commented on lib/gpprmsya.gi:
```diff
@@ -349,51 +349,57 @@ local   F,      # free group
+local   F,      # free group
+	gens,	#generators of F
+	imgs,
+	hom,	# bijection
```
Done.

hulpke replied:
I have replaced tabs.

hulpke commented on lib/gpprmsya.gi:
```diff
@@ -349,51 +349,57 @@ local   F,      # free group
+	i, j;       # loop variables
+
+    mov:=MovedPoints(G);
+    deg:=Length(mov);
```
Changed

hulpke commented on lib/gpprmsya.gi:
```diff
@@ -349,51 +349,57 @@ local   F,      # free group
+    od;
+
+    # return the isomorphism to the finitely presented group
+    hom:= GroupHomomorphismByImages(G,F,imgs,GeneratorsOfGroup(F));
```
Done

hulpke replied:
No, `gens` are free generators, while `F` now is fp.

Commit:
ENHANCE: Improvements to IrreducibleModules

Burnside-Brauer method for permutation groups.

Generically:
Factor out radical part stemming from P-core.
Spin submodules spanned by augmentation ideal vectors for first submodul
chain and chop, to avoid irreducibility tests in large regular module.

Also flag fix for `TrivialModule`.

Commit:
ENHANCE: CompatiblePairs also works if group is not solvable

Uses Module isomorphism/automorphism code, modeled on Cannon/Holt autom.
group paper.

Also add action of compatible pairs

Commit:
CLEANUP: Replace tabs with blanks as requested

fingolfin commented on lib/twocohom.gi:
```diff
@@ -691,7 +853,621 @@ function( G, M )
+          AddSet(hastail,Length(rules));
+      fi;
+    elif Length(r[1])>1 then
+      if Length(r[2])=0 then Error("generator is trivial");fi;
```
Also not covered by `testinstall`.

fingolfin replied:
Still not covered.

fingolfin commented on lib/twocohom.gi:
```diff
@@ -691,7 +853,621 @@ function( G, M )
+  if model<>fail then
+    q:=GQuotients(model,G)[1];
+    pre:=List(gens,x->PreImagesRepresentative(q,x));
+    ker:=KernelOfMultiplicativeGeneralMapping(q);
```
Also not covered by `testinstall`.

fingolfin commented on lib/twocohom.gi:
```diff
@@ -691,7 +853,621 @@ function( G, M )
+  r.pairact:=function(zy,pair)
+  local autom,mat,i,imagemonwords,imgwrd,left,right,extim,prdout,myinvers,v;
+
+    autom:=InverseGeneralMapping(pair[1]);
```
Also not covered by `testinstall`.

fingolfin replied:
This is now covered, nice!

fingolfin commented on lib/twocohom.gi:
```diff
@@ -691,7 +853,621 @@ function( G, M )
+
+InstallGlobalFunction("CompatiblePairOrbitRepsGeneric",function(cp,coh)
+local bas,ran,mats,o;
+  if Length(coh.cohomology)=0 then return [coh.zero];fi;
```
Also not covered by `testinstall`.

alex-konovalov commented on lib/grppcext.gi:
```diff
@@ -497,13 +497,58 @@ end);
+    elif Length(arg)=2 then
+      # search through automorphism group for projection image and reps,
+      # then add module automorphisms
+      gens:=GeneratorsOfGroup(G);
```
I could run the test of this PR in Jenkins, to exercise testextra. However, it does not pass cleanly at the moment: #3408

hulpke replied:
@fingolfin I have added smaller a test to `testinstall' that checks for action of compatible pairs (thus testing most code). Is seems to run through (the current test failure is due to a DNS resolving problem).

fingolfin replied:
Thanks. I'll go over the coverage once again.

fingolfin commented on lib/grppcext.gi:
```diff
@@ -497,13 +497,58 @@ end);
+      B:=MTX.ModuleAutomorphisms(M);
+      if Size(B)>1 then
+        for i in Set(GeneratorsOfGroup(B)) do
+          Add(pool,DirectProductElement([One(A),i]));
```
This loop is still not covered.

fingolfin commented on lib/grppcext.gi:
```diff
@@ -497,13 +497,58 @@ end);
+
+    if oper=fail then
+      Ggens:=GeneratorsOfGroup(G);
+      oper := GroupHomomorphismByImagesNC( G, Mgrp, Ggens, M.generators );
```
Still not covered.

fingolfin commented on lib/grppcext.gi:
```diff
@@ -635,11 +681,27 @@ local G, M, Mgrp, oper, A, B, D, translate, gens, genimgs, triso, K, K1,
-    else
+    if not IsPcgs(Ggens) then
+      elmlist:=fail;
+      epi:=EpimorphismFromFreeGroup(G);
```
Still not covered.

fingolfin commented on lib/grppcext.gi:
```diff
@@ -648,151 +710,157 @@ local G, M, Mgrp, oper, A, B, D, translate, gens, genimgs, triso, K, K1,
+      until l=1;
+    else
+      #D:=Stabilizer(D,M.generators,gens,genimgs,f);
+      hf:=SparseIntKeyVecListAndMatrix(false,Concatenation(M.generators));
```
Still not covered.

fingolfin commented on lib/grpreps.gi:
```diff
@@ -44,21 +44,195 @@ end);
+    else
+      a:=MaximalAbelianQuotient(G);
+      si:=List(gens,x->ImagesRepresentative(a,x));
+      sub:=IrreducibleModules(Group(si),F,1);
```
Still not covered.

fingolfin commented on lib/grpreps.gi:
```diff
@@ -44,21 +44,195 @@ end);
+
+  # The augmentation ideal of a P-normal subgroup lies in the radical
+  if Characteristic(F)=0 then
+    si:=TrivialSubgroup(G);
```
Still not covered.

fingolfin commented on lib/grpreps.gi:
```diff
@@ -44,21 +44,195 @@ end);
+    ser:=[[],subs[1],One(mats[1])];
+    for i in [2..Length(subs)] do
+      a:=subs[i];
+      j:=2;
```
Still not covered.

fingolfin commented on lib/twocohom.gi:
```diff
@@ -667,6 +667,168 @@ function( G, M )
+    monreps:=fmgens{[loff+1..off]};
+    monreal:=mgens{[loff+1..off]};
+    if IsBound(m!.rewritingSystem) then
+      k:=m!.rewritingSystem;
```
This is one of only two lines in this function which is no covered. Very nice!

hulpke replied:
Yes, this is as a hook for future extensions (RWS for simple groups).

fingolfin commented on lib/twocohom.gi:
```diff
@@ -667,6 +667,168 @@ function( G, M )
+            else x:=-i/2;fi;
+            # word must be freely cancelled
+            if Length(g)>0 and x=-g[Length(g)] then
+              Unbind(g[Length(g)]);
```
(This is the other uncovered line).

fingolfin commented on lib/twocohom.gi:
```diff
@@ -691,7 +853,621 @@ function( G, M )
+
+    # collect from left
+    i:=1;
+    while i<=Length(a) do
```
This is now covered, thanks.

fingolfin reviewed (approved):
Thank you for the added `testinstall` test. There are still about 300 new lines of code which are not covered by a test in `testinstall`, and the new code only has a coverage rate of ~76% as a result. I'd prefer if it was higher (not because high coverage by itself is a magic goal, but rather because it would increase my confidence in all those special cases, and also in our ability to keep this code from breaking in the future).

Anyway, I don't think we need to delay this further because of that. So let's merge it soon. (I'll wait for Travis to finish with its current truck loads of builds before merging it).
